VMEncryption/volume-notif-svc/UdevVolNotif.c

```diff
@@ -0,0 +1,471 @@
+        custom_log("ERROR: Can't fork (error no=%d)\n", errno);
+        exit(1);
+    }
+    if (pid > 0) {
+        custom_log("Parent process is exiting, pid=%d", pid);
+        exit(0);
+    }
+    umask(0);
+    pid_t sid = setsid();
+    if (sid < 0) {
+        custom_log("ERROR: Can't set sid (error no=%d)\n", errno);
+        exit(1);
+    }
+    if ((chdir("/")) < 0) {
+        custom_log("ERROR: Can't change directory (error no=%d)\n", errno);
+        exit(1);
+    }
+    close(STDIN_FILENO);
+    close(STDOUT_FILENO);
+    close(STDERR_FILENO);
+
+    open("/dev/null",O_RDWR);
+    dup(0);
+}
+
+void invoke_ade(const char* path,std::deque<device>&dq){
+    ade_status = AdeStatus::ADE_RUNNING;
+    //update counter in dq
+    for(auto item = dq.begin(); item!=dq.end(); item++){
+       item->ade_attempt_count+=1;
+    }
+    pid_t pAde = fork();
+    if(pAde==0){
+            // in case of child process.
+            custom_log("child process %d created, running ADE", getpid());
+            custom_log("changing directory to %s", path);
+            chdir(path);
+            execl(          "extension_shim.sh",
+                            "extension_shim.sh",
+                            "-c",
+                            "main/handle.py --enable --vnscall",
+                            NULL);
+            int err = errno;
+            custom_log("Failed to run process! error: %d",err);
+            exit(1);
+    }
+    if(pAde>0){
+       pid_t childPid = wait(NULL);
+       custom_log("child process %d is completed",childPid);
+    }
+    //checking if daemon is still running
+    pid_t pDaemon = fork();
+    if(pDaemon==0){
+            custom_log("child process %d created, checking daemon", getpid());
+            custom_log("changing directory to %s", path);
+            chdir(path);
+            execl(          "ade_daemon_delay.sh",
+                            "ade_daemon_delay.sh",
+                            "extension_shim.sh",
+                            NULL);
+            int err = errno;
+            custom_log("Failed to run process! error: %d",err);
+            exit(1);
+
+    }
+    if(pDaemon>0){
+       pid_t childPid = wait(NULL);
+       custom_log("child process %d is completed",childPid);
+    }
+    ade_status = AdeStatus::ADE_FINISHED;
+    ade_finished = time(NULL);
+}
+
+int is_device_crypted_from_syspath(const char* syspath){
+        struct udev *udev;
+        int ret = 0;
+        udev = udev_new();
+        struct udev_device* dev = udev_device_new_from_syspath(udev,syspath);
+        const char* fs_usage = udev_device_get_property_value(dev,"ID_FS_USAGE");
+        if(fs_usage!=NULL && strcmp(fs_usage,"crypto")==0) ret = 1;
+        udev_device_unref(dev);
+        udev_unref(udev);
+        return ret;
+}
+
+bool is_devnode_added(std::deque<device>&dq, const char* syspath, const char* action){
+        for(auto it:dq)
+                if(             strcmp(it.syspath.c_str(),syspath)==0 &&
```

Comment:
In a subsequent, non-edit PR, run document formatter for fixing the syntax consistency issues.